### PR TITLE
chore(wallet): prepare swap UI to handle LiFi same-chain swaps

### PR DESCRIPTION
**Pull request by josheleonard**
Resolves https://github.com/brave/brave-browser/issues/37212

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/swap_service.cc
Comment (StephenHeaps) [inline components/brave_wallet/browser/swap_service.cc:1]: Adjusting `GetQuote` locally to try LiFi first, I hit a scenario locally where I received a `LiFiQuote` with 0 routes on Solana mainnet (0.1 SOL -> USDC), should/could we fallback to Jupiter (or ZeroX) internally in the case of 0 available routes? cc @onyb 

May be Solana specific; I haven't hit this on a network supported by LiFi but not ZeroX so current priorities in `GetQuote` might make this irrelevant.

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 1]
```diff
@@ -777,6 +836,15 @@ export const useSwap = () => {
         return 'insufficientLiquidity'
       }
 
+      // LiFi specific validations
+      if (quoteErrorUnion?.lifiError?.message) {
+        return quoteErrorUnion?.lifiError?.message?.includes(
+          'Could not find token'
+        )
+          ? 'insufficientLiquidity'
+          : 'unknownError'
```
Comment (StephenHeaps) [inline components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts:845]: I wonder if this is better moved into core on the `LiFiError` object itself similar to `JupiterError` & `ZeroExError`? I wasn't able to find insufficient liquidity in a few quick tests but I _believe_ we can check for [`INSUFFICIENT_LIQUIDITY`](https://docs.li.fi/li.fi-api/li.fi-api/requesting-a-quote/possible-tool-errors) code, or move this string includes/contains check into core. Then we wouldn't have to update frontend if we added code support later.

Comment (StephenHeaps) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts]: Awesome thanks! Updated iOS side in https://github.com/brave/brave-core/pull/22652/commits/0ed8b521eefc11b0bab7535a09aa7dc629c307b4 to check the code enum too.

File: components/brave_wallet/common/brave_wallet.mojom [hunk 2]
```diff
@@ -738,8 +739,59 @@ struct LiFiEVMTransaction {
...
+  NotFoundError = 1003,
+  NotProcessableError = 1004,
+  RateLimitError = 1005,
+  ServerError = 1006,
+  SlippageError = 1007,
+  ThirdPartyError = 1008,
+  TimeoutError = 1009,
+  UnauthorizedError = 1010,
+  ValidationError = 1011,
+};
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet.mojom:783]: Guide requires us to use kPrefxied convention.
So `LifiToolErrorCode.kNoPossbileRoute`, `LiFiErrorCode.kDefaultError` and so on
https://chromium.googlesource.com/chromium/src/+/master/mojo/public/tools/bindings/README.md#enumeration-types

File: components/brave_wallet/common/brave_wallet.mojom [hunk 3]
```diff
@@ -738,8 +739,59 @@ struct LiFiEVMTransaction {
...
+  UNKNOWN_ERROR = 3,
+  // There was a problem getting on-chain data. Please try again later.
+  RPC_ERROR = 3,
+  // he initial amount is too low to transfer using this tool.
+  AMOUNT_TOO_LOW = 4,
+  // The initial amount is too high to transfer using this tool.
+  AMOUNT_TOO_HIGH = 5,
+  // The fees are higher than the initial amount.
+  // This would result in negative resulting token.
+  FEES_HGHER_THAN_AMOUNT = 6,
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet.mojom:759]: Higher

File: components/brave_wallet/browser/swap_service_unittest.cc [hunk 4]
```diff
@@ -201,7 +202,12 @@ mojom::LiFiQuotePtr GetCannedLiFiQuote() {
   fee_cost->amount = "853380";
   fee_cost->included = true;
   fee_cost->token = from_token.Clone();
-  quote->routes[0]->steps[0]->estimate->fee_costs.push_back(fee_cost.Clone());
+
+  if (!quote->routes[0]->steps[0]->estimate->fee_costs) {
+  } else {
+    quote->routes[0]->steps[0]->estimate->fee_costs->push_back(
+        fee_cost.Clone());
+  }
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_service_unittest.cc:211]:
```
  if (quote->routes[0]->steps[0]->estimate->fee_costs) {
    quote->routes[0]->steps[0]->estimate->fee_costs->push_back(
        fee_cost.Clone());
  }
```

Comment (supermassive) [reply components/brave_wallet/browser/swap_service_unittest.cc:211]: and same below

File: components/brave_wallet/browser/swap_service_unittest.cc [hunk 5]
```diff
@@ -1450,10 +1459,14 @@ TEST_F(SwapServiceUnitTest, GetLiFiQuoteError) {
...
+
+  auto lifi_error = mojom::LiFiError::New();
+  lifi_error->message = "Invalid request";
+
   EXPECT_CALL(callback, Run(EqualsMojo(mojom::SwapQuoteUnionPtr()),
                             EqualsMojo(mojom::SwapFeesPtr()),
                             EqualsMojo(mojom::SwapErrorUnion::NewLifiError(
-                                mojom::LiFiError::New("Invalid request"))),
+                                std::move(lifi_error))),
                             ""));
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_service_unittest.cc:1470]: inplace construction with expected error code should work 
`mojom::LiFiError::New("Invalid request", mojom::LiFiErrorCode::kSomethingExpected)))`
(also for next test)

File: components/brave_wallet/browser/swap_service.cc [hunk 6]
```diff
@@ -521,6 +524,17 @@ void SwapService::OnGetLiFiQuote(mojom::SwapFeesPtr swap_fee,
...
 
   if (auto quote = lifi::ParseQuoteResponse(api_request_result.value_body())) {
+    if (quote->routes.empty()) {
+      auto error_response = mojom::LiFiError::New();
+      error_response->code = mojom::LiFiErrorCode::NotFoundError;
+      error_response->message = "No Routes found";
+
+      std::move(callback).Run(
+          nullptr, nullptr,
+          mojom::SwapErrorUnion::NewLifiError(std::move(error_response)), "");
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_service.cc:534]: prefer inplace construction
```
mojom::SwapErrorUnion::NewLifiError(mojom::LiFiError::New("No Routes found", mojom::LiFiErrorCode::NotFoundError))
```

File: components/brave_wallet/browser/swap_responses.idl [hunk 7]
```diff
@@ -111,7 +111,7 @@ namespace swap_responses {
     // The following fields are mandatory for EVM -> any transfers.
     DOMString? from;
     DOMString? to;
-    DOMString? chainId;
+    long? chainId;
     DOMString? value;
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_responses.idl:115]: How did this work?
Do/should we use `ConvertAllNumbersToString` converter so all numbers appear as strings?

File: components/brave_wallet/browser/swap_response_parser.cc [hunk 8]
```diff
@@ -388,9 +390,12 @@ mojom::BlockchainTokenPtr ParseToken(const swap_responses::LiFiToken& value) {
   auto result = mojom::BlockchainToken::New();
   result->name = value.name;
   result->symbol = value.symbol;
-  result->logo = value.logo_uri;
+  result->logo = value.logo_uri.has_value() ? value.logo_uri.value() : "";
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_response_parser.cc:393]: `.value_or("")`

File: components/brave_wallet/browser/swap_response_parser.cc [hunk 9]
```diff
@@ -388,9 +390,12 @@ mojom::BlockchainTokenPtr ParseToken(const swap_responses::LiFiToken& value) {
...
   result->contract_address =
-      value.address == kLiFiNativeAssetContractAddress ? "" : value.address;
+      value.address == kLiFiNativeEVMAssetContractAddress ||
+              value.address == kLiFiNativeSVMAssetContractAddress
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_response_parser.cc:396]: I'd wrap it like this for clarity:
`(value.address == kLiFiNativeEVMAssetContractAddress || value.address == kLiFiNativeSVMAssetContractAddress)`

File: components/brave_wallet/browser/swap_response_parser.cc [hunk 10]
```diff
@@ -472,7 +471,22 @@ mojom::LiFiStepEstimatePtr ParseEstimate(
...
+    auto fee_cost = mojom::LiFiFeeCost::New();
+    fee_cost->name = fee_cost_value.name;
+    fee_cost->description = fee_cost_value.description;
+    fee_cost->percentage = fee_cost_value.percentage;
+    fee_cost->token = ParseToken(fee_cost_value.token);
+    fee_cost->amount = fee_cost_value.amount;
+    fee_cost->included = fee_cost_value.included;
+    fee_costs.push_back(std::move(fee_cost));
+  }
+  result->fee_costs = std::move(fee_costs);
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_response_parser.cc:489]: Let's not do early return here. Someone might add some parsing after this block which would be incorrectly lost when there is no fee_costs
So this way is preferred
```
  if (value.fee_costs) {
    for (const auto& fee_cost_value : *value.fee_costs) {
      ...
      result->fee_costs.push_back(std::move(fee_cost));
    }
  }
```

File: components/brave_wallet/browser/swap_request_helper.cc [hunk 11]
```diff
@@ -307,6 +303,32 @@ std::optional<base::Value::Dict> EncodeStepEstimate(
...
+    fee_cost_value.Set("name", fee_cost->name);
+    fee_cost_value.Set("description", fee_cost->description);
+    fee_cost_value.Set("amount", fee_cost->amount);
+    fee_cost_value.Set("percentage", fee_cost->percentage);
+    fee_cost_value.Set("included", fee_cost->included);
+
+    // fake the USD amount value since it is not used by LiFi
+    fee_cost_value.Set("amountUSD", "0");
+
+    if (auto token = EncodeToken(std::move(fee_cost->token))) {
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_request_helper.cc:322]: fee_cost is const so std::move has no effect. Actually all calls to EncodeToken shold not recieve args by std::move

File: components/brave_wallet/common/brave_wallet.mojom (hunk 2 above)
Comment (josheleonard) [reply components/brave_wallet/common/brave_wallet.mojom]: fixed

File: components/brave_wallet/common/brave_wallet.mojom (hunk 3 above)
Comment (josheleonard) [reply components/brave_wallet/common/brave_wallet.mojom]: thanks, fixed.

File: components/brave_wallet/browser/swap_service_unittest.cc (hunk 4 above)
Comment (josheleonard) [reply components/brave_wallet/browser/swap_service_unittest.cc]: fixed

File: components/brave_wallet/browser/swap_response_parser.cc (hunk 8 above)
Comment (josheleonard) [reply components/brave_wallet/browser/swap_response_parser.cc]: fixed

File: components/brave_wallet/browser/swap_service_unittest.cc (hunk 5 above)
Comment (josheleonard) [reply components/brave_wallet/browser/swap_service_unittest.cc:1468]: updated, but the test still fails

File: components/brave_wallet/browser/swap_response_parser.cc (hunk 9 above)
Comment (josheleonard) [reply components/brave_wallet/browser/swap_response_parser.cc]: wrapped

File: components/brave_wallet/browser/swap_service.cc (hunk 6 above)
Comment (josheleonard) [reply components/brave_wallet/browser/swap_service.cc:529]: inlined

**Comment by supermassive** [comment]
@josheleonard 
I went ahead ahead and fixed some tests in https://github.com/brave/brave-core/pull/22652/commits/1f82d6b3f95510cfa3e30e34a7a3a8793d1335af
PTAL
- `chainId` is a string again
- Reverted SwapService::OnGetLiFiQuote: should handle lifi errors and possible parsing errors both.
- There are some tests still fail. They need adding new fields (amountUSD, priceUSD, logoURI, etc) in expected jsons where acceptable.

**Review by supermassive — APPROVED** [review]
wallet core lgtm

**Review by onyb — CHANGES_REQUESTED** [review]

File: components/brave_wallet/browser/swap_responses.idl (hunk 7 above)
Comment (onyb) [reply components/brave_wallet/browser/swap_responses.idl]: `ConvertAllNumbersToString` appears to be used now in `swap_service.cc` so lgtm.

File: components/brave_wallet/common/brave_wallet.mojom [hunk 12]
```diff
@@ -738,8 +739,59 @@ struct LiFiEVMTransaction {
   string gas_limit;
 };
 
+enum LifiToolErrorCode {
+  // No route was found for this action.
+  NO_POSSIBLE_ROUTE = 0,
```
[the remark this answers is not in the document]
Comment (josheleonard) [reply components/brave_wallet/common/brave_wallet.mojom]: fixed: [aefdba9](https://github.com/brave/brave-core/pull/22652/commits/aefdba9853670c2a892a57370736cac424d2d22a)

File: components/brave_wallet/common/brave_wallet.mojom [hunk 13]
```diff
@@ -738,8 +739,59 @@ struct LiFiEVMTransaction {
...
+  kServerError = 1006,
+  kSlippageError = 1007,
+  kThirdPartyError = 1008,
+  kTimeoutError = 1009,
+  kUnauthorizedError = 1010,
+  kValidationError = 1011,
+};
+
+struct LifiToolError {
+  string errorType;
```
[the remark this answers is not in the document]
Comment (josheleonard) [reply components/brave_wallet/common/brave_wallet.mojom]: fixed: [aefdba9](https://github.com/brave/brave-core/pull/22652/commits/aefdba9853670c2a892a57370736cac424d2d22a)

File: components/brave_wallet/browser/swap_service.cc [hunk 14]
```diff
@@ -521,6 +524,17 @@ void SwapService::OnGetLiFiQuote(mojom::SwapFeesPtr swap_fee,
   }
 
   if (auto quote = lifi::ParseQuoteResponse(api_request_result.value_body())) {
+    if (quote->routes.empty()) {
+      auto error_response = mojom::LiFiError::New();
+      error_response->code = mojom::LiFiErrorCode::kNotFoundError;
+      error_response->message = "No Routes found";
```
[the remark this answers is not in the document]
Comment (josheleonard) [reply components/brave_wallet/browser/swap_service.cc]: internationalized: [09ace8c](https://github.com/brave/brave-core/pull/22652/commits/09ace8ccb2bb5b130d57bab9c47820ce3ef3e0dc)

**Review by onyb — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/page/screens/swap/hooks/useLifi.ts [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+  // methods
+  const exchange = useCallback(
+    async function (step: BraveWallet.LiFiStep) {
+      if (
+        !fromAccount ||
+        !toAccountId ||
+        !fromToken ||
+        !toToken ||
+        !fromNetwork ||
+        !step
```
[the remark this answers is not in the document]
Comment (josheleonard) [reply components/brave_wallet_ui/page/screens/swap/hooks/useLifi.ts]: removed: [dc83066](https://github.com/brave/brave-core/pull/22652/commits/dc83066eaa0999c71d45dc67266798156ab7330a)

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 16]
```diff
@@ -412,6 +443,34 @@ export const useSwap = () => {
...
+          if (params.editingFromOrToAmount === 'from') {
+            setToAmount(getLiFiToAmount(route).format(6))
+          } else {
+            setFromAmount(getLiFiFromAmount(route).format(6))
+          }
+
+          const step = route.steps[0]
+
+          if (step) {
+            await checkAllowance({
```
[the remark this answers is not in the document]
Comment (josheleonard) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts:465]: Good catch! fixed: [dc83066](https://github.com/brave/brave-core/pull/22652/commits/dc83066eaa0999c71d45dc67266798156ab7330a)

**Review by onyb — APPROVED** [review]
LGTM

File: ios/brave-ios/Sources/BraveWallet/Crypto/WalletActions/SwapCryptoView.swift [hunk 17]
```diff
@@ -241,8 +241,8 @@ struct SwapCryptoView: View {
 
   /// The DEX Aggregator for the current network.
   var dexAggregator: DEXAggregator {
-    // TODO(stephenheaps): Once LiFiQuote is supported, we need to
-    // update this disclaimer to include LiFi description & privacy
+    // TODO(stephenheaps): We may need to remove and/or update
+    // this disclaimer to include LiFi description & privacy
```
Comment (nuo-xu) [inline ios/brave-ios/Sources/BraveWallet/Crypto/WalletActions/SwapCryptoView.swift:245]: maybe a link to the slack thread discussing the update of the disclaimer copy.

**Review by nuo-xu — APPROVED** [review]
iOS++

File: ios/brave-ios/Sources/BraveWallet/Crypto/WalletActions/SwapCryptoView.swift (hunk 17 above)
Comment (StephenHeaps) [reply ios/brave-ios/Sources/BraveWallet/Crypto/WalletActions/SwapCryptoView.swift:245]: We're waiting on some info for this change: https://bravesoftware.slack.com/archives/C023VS4HJ6Q/p1711451893091189?thread_ts=1708974083.632369&cid=C023VS4HJ6Q

FWIW LiFi is not yet enabled in this PR.

File: components/brave_wallet/browser/swap_responses.idl (hunk 7 above)
Comment (StephenHeaps) [reply components/brave_wallet/browser/swap_responses.idl]: This change from `long?` to `DOMString?` is causing failed to parse error when creating a transaction from a `LiFiQuote` now. Reverting back to `long? chainId` resolves parsing failure. Might be `ConvertAllNumbersToString` not converting for optional types?

**Review by StephenHeaps — CHANGES_REQUESTED** [review]
See inline comment; failed to parse response error when creating a transaction.

File: components/brave_wallet/browser/swap_responses.idl (hunk 7 above)
Comment (josheleonard) [reply components/brave_wallet/browser/swap_responses.idl]: changing back to from `DOMString` to `long?` causes unit tests to fail

Comment (josheleonard) [reply components/brave_wallet/browser/swap_responses.idl]: Thoughts @supermassive , @onyb ?

Comment (StephenHeaps) [reply components/brave_wallet/browser/swap_responses.idl:115]: Fixed with [cb383f6](https://github.com/brave/brave-core/pull/22652/commits/cb383f6f220ad7944a523cf5bdab7912b20fabba)! 🐎

**Review by StephenHeaps — APPROVED** [review]

**Review by mkarolin — APPROVED** [review]
`strings`++

**Pull request merged**